File: Array/second_largest.py
# O(n+n)
import sys
def getSecondOrderElements(n,a) -> [int]:
    max_element = a[0]
    min_element = a[0]
    for i in range(n):
        if a[i]> max_element:
            max_element = a[i]

        if a[i]< min_element:
            min_element = a[i]

    pre_max = -sys.maxsize
    pre_min = sys.maxsize

    for i in range(n):
        if a[i]> pre_max and a[i]< max_element:
            pre_max = a[i]

        if a[i]< pre_min and a[i]> min_element:
            pre_min = a[i]

    return [pre_max, pre_min]



#Optimal Solution
def getSecondOrderElements2(n,a):
    max_ele = a[0]
    second_max = 0

    for i in range(n):
        if a[i]>max_ele:
            max_ele = a[i]
            second_max = 


print("Better")
print(getSecondOrderElements(6,[2,3,4,5,9,8]))
print("Optimal")
print(getSecondOrderElements2(6,[2,3,4,5,9,8]))

File: Array/second_largest.cpp
#include <vector>
#include <iostream>
#include <climits>
using namespace std;

vector<int> getSecondOrderElements(int n, vector<int> a) {
    int max_element = a[0];
    int min_element = a[0];

    for(int i=1; i<n; i++){
        if (a[i]> max_element){
            max_element = a[i];
        }

        if (a[i]< min_element){
            min_element = a[i];
        }

    }

    int pre_max = INT_MIN;
    int pre_min = INT_MAX;

    for (int i=0; i<n; i++){
        if (a[i] > pre_max && a[i] < max_element){
            pre_max = a[i];
        }

        if (a[i]< pre_min && a[i]> min_element){
            pre_min = a[i];
        }

    }

    vector<int> result;
    result.push_back(pre_max);
    result.push_back(pre_min);

    return result;

}

int main() {
    int n = 5;
    vector<int> a = {1, 2, 3, 4, 5};

    vector<int> result = getSecondOrderElements(n, a);
    cout << result[0] << " " << result[1] << endl;
    return 0;
}